Jael rewrite ("Kale")

philipcmonk opened the pull request:
This finishes Kale and replaces Jael with it throughout the system.  "Kale" was a codename, so everything is now named back to "Jael".  Changes:

- Handle reorgs correctly.  This means we check each new block to make sure its parent hash matches the last one we saw, and we rewind if it doesn't.  We also wait 30 blocks before considering an event "real" to avoid flapping on transient reorgs.

- Split the interaction with an Ethereum node into a separate app `azimuth-tracker`.

- Reworked moons.  Moons are no longer self-attesting.  Instead, when someone needs their keys, they subscribe to the moon's parent for their keys.  The parent is responsible for pushing out key updates in case the moon breaches.

- Final Jael support for individual breaches, specifically regarding the case where your sponsor breaches without warning.  This case is similar to bootstrapping and must be handled the same way.  How to determine the true state of the PKI without previous state?  Proof-of-work is the solution to this problem, so we listen directly to an Ethereum node.  Thus, you generally get PKI updates in the following ways:
  - Updates about your sponsor come from an Ethereum node
  - Updates about any other planet, star, or galaxy come from your sponsor.
  - Updates about a moon come from that moon's parent.
  - Comets cannot have updates, as they are self-attesting.

- Lines of code greatly reduced:
  - Old Jael: over 2500
  - New Jael: about 1150
  - New Azimuth-tracker: about 400

jtobin commented on pkg/arvo/app/acme.hoon:
```diff
@@ -344,7 +344,7 @@
 =/  directory-base=purl
   =-  (need (de-purl:html -))
-  'https://acme-v02.api.letsencrypt.org/directory'
+  'https://acme-staging-v02.api.letsencrypt.org/directory'
```
Guessing this staging URL might have slipped in unnoticed.

philipcmonk replied:
yep, fixed.

belisarius222 commented on pkg/arvo/app/azimuth-tracker.hoon:
```diff
@@ -40,13 +38,16 @@
+          ==
+      ?:  =(~ ships)
+        ~
+      [(turn ~(tap in ships) ,@) ~]
```
The last few lines of this function look a bit funny.  A return type annotation would be helpful.

philipcmonk replied:
Added return type and a couple comments

belisarius222 commented on pkg/urbit/daemon/main.c:
```diff
@@ -327,8 +327,8 @@ _main_getopt(c3_i argc, c3_c** argv)
-    u3_Host.ops_u.sap_c =
-        "https://bootstrap.urbit.org/urbit-" URBIT_VERSION ".snap";
+    u3_Host.ops_u.sap_c = 0;
+        // "https://bootstrap.urbit.org/urbit-" URBIT_VERSION ".snap";
```
looks like a debugging artifact

philipcmonk replied:
Yep, related to removing snapshots.

belisarius222 commented on pkg/arvo/sys/vane/jael.hoon:
```diff
@@ -1097,1181 +604,310 @@
+      this-su
+    (public-keys:feel %diff a-ship u.a-diff)
+  ::
+  ++  extract-snap                                      ::  extract rewind point
```
It looks like snapshots are no longer operational, but there are a couple stubs like this one left.  Can you explain how you see this going in the future?

philipcmonk replied:
That's correct.  Now that we only listen to key changes, sponsor changes, and breaches, It seems like they're not needed, at least for now.  On mainnet it takes less than a minute of processing time to get up to date, and it doesn't block for that whole time.  It's faster on ropsten, and of course fake ships don't take any time for this.

So I've removed snapshots for now to reduce complexity.  We might decide to add them back, and we can do so much more cleanly now.

belisarius222 commented:
It feels a little weird to leave in pieces of the snapshot machinery.  I wonder if we should rip out the atavistic functions and command-line arguments related to snapshots.

—
~rovnys-ricfer
https://urbit.org

On Wed, Aug 07, 2019 at 10:44 AM, Philip Monk < notifications@github.com > wrote:

> 
> 
> 
> *@philipcmonk* commented on this pull request.
> 
> 
> 
> In pkg/ arvo/ app/ acme. hoon (
> https://github.com/urbit/urbit/pull/1436#discussion_r311680950 ) :
> 
> > @@ -344,7 +344,7 @@ :: =/ directory-base=purl =- (need (de-purl:html -))
> - ' https:/ / acme-v02. api. letsencrypt. org/ directory (
> https://acme-v02.api.letsencrypt.org/directory ) ' + ' https:/ / acme-staging-v02.
> api. letsencrypt. org/ directory (
> https://acme-staging-v02.api.letsencrypt.org/directory ) '
> 
> yep, fixed.
> 
> 
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub (
> https://github.com/urbit/urbit/pull/1436?email_source=notifications&email_token=AAGVR5KGQDGQZ6H7JQPZC2LQDMCY3A5CNFSM4IJ3BO6KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCA4E3NY#discussion_r311680950
> ) , or mute the thread (
> https://github.com/notifications/unsubscribe-auth/AAGVR5JVVYJD7YMOBBAAUMDQDMCY3ANCNFSM4IJ3BO6A
> ).
> 
> 
>

joemfb commented on pkg/arvo/sys/vane/jael.hoon:
```diff
@@ -2416,55 +1047,50 @@
-    ``[%atom !>(sig)]
+    =/  sec  (~(got by jaw.own.pki.lex) u.lyf)
+    =/  moon-sec  (shaf %earl (sham our u.lyf sec u.who))
+    =/  cub  (pit:nu:crub:crypto 128 moon-sec)
```
Do we gain anything (other than scry convenience) by deterministically deriving the moon keys from that parent's? If peers have to ask us for the moon's public key anyway, wouldn't it be better to just statefully derive the moon's keys from entropy?

philipcmonk replied:
Yes, they should be statefully derived.  Basically, there should be a `|moon` command that registers a public key with the planet's jael.  If no key is supplied, it should generate and output a private key.

This is all formally correct for now, but I'll try to build |moon soon.

philipcmonk replied:
Done, PTAL at gen/hood/moon{,-breach,-cycle-keys}.hoon

joemfb commented on pkg/arvo/lib/hood/helm.hoon:
```diff
@@ -86,7 +85,8 @@
   =/  rit  (sy [%jewel (my [lyf.u.sed key.u.sed] ~)] ~)
-  (emit %mint / our rit)
+  ~&  %rekey-not-implemented
+  +>.$
```
Is this just triage, or is there some obstacle to reimplementing re-key a la `%mint`?

philipcmonk replied:
I didn't notice that %mint existed until I was cleaning something up right at the end, so I didn't realize that rekey was possible.  It shouldn't be hard to re-add.

philipcmonk replied:
Done

joemfb commented:
In a similar vein to the snapshot considerations, it appears that nothing using the `(unit oauth)` signature in `+deed:able:ames` and `+seed:able:jael`. Is there any reason to keep that around?

philipcmonk commented:
Currently comets use the oath for self attestation.  I think that will change in alef, so it should be removed there.

I think you're right we may as well rip out the snapshot machinery altogether, I'll do that.

philipcmonk commented on nix/ops/solid/builder.sh:
```diff
@@ -19,6 +19,11 @@ trap cleanup EXIT
+until [ -d ./pier/home ]
+do
+  sleep 1
+done
```
I guess?  I only added this because it was a problem.  I don't know how defensively we want to program here.

philipcmonk replied:
added

philipcmonk commented on pkg/arvo/gen/solid.hoon:
```diff
@@ -15,7 +15,8 @@
     ==
-:-  %noun
+:-  %pill
+%-  pill:pill
```
The reason to make `+solid` produce a well-typed pill is that we now have circumstances where it's reasonable to write it to clay or pass it into aqua.  This lets you write `:aqua +solid` instead of `:aqua &pill +solid`.

I haven't noticed any performance impact, the pill type isn't very specific, so the clam doesn't need to do much.  However, I'm pretty sure I can just change `+file-ovum` to produce a `unix-event`, then it can be a cast.

philipcmonk commented on pkg/arvo/sys/zuse.hoon:
```diff
@@ -7887,6 +7619,7 @@
         ::  launch: block number of azimuth deploy
         ::
         ++  launch  6.784.800
+        ::  ++  launch  4.601.630
```
If we have the ropsten contract address a few lines above, then we should have the ropsten block number here.  Not sure what's the cleanest way to do this.

philipcmonk replied:
I'm just going to define a `+mainnet-contracts` and ` +ropsten-contracts`, then let `++  contracts  mainnet-contracts`.  Switching to ropsten just means switching that line.

philipcmonk commented on pkg/arvo/sys/vane/ames.hoon:
```diff
@@ -1386,7 +1369,7 @@
       ::  ~&  [%milk p.bon q.bon]
       ?>  ?=([@ @ *] q.q.bon)
-      ?>  ?=(?(%a %c %e %g %j) i.q.q.bon)
+      ?>  ?=(?(%a %c %e %g %j %j) i.q.q.bon)
```
fixed

jtobin commented:
On Wed, Aug 07, 2019 at 01:35:14PM -0700, Philip Monk wrote:
> I'm just going to define a `+mainnet-contracts` and `
> +ropsten-contracts`, then let `++  contracts  mainnet-contracts`.
> Switching to ropsten just means switching that line.

As a side note, I'd like to sooner-than-later make this something that
can be toggled via a command-line option in vere, e.g. by way of
'--testnet-mode' or what have you.  Having to actually edit the vane,
etc. to spin up a testnet is a real pain.

joemfb commented on pkg/urbit/daemon/main.c:
```diff
@@ -164,10 +163,6 @@ _main_getopt(c3_i argc, c3_c** argv)
         break;
       }
-      case 'm': {
-        u3_Host.ops_u.sap_c = strdup(optarg);
```
Please also remove `sap_c` and `etn` from `u3_opts`. (I know that struct is already out of date, but I don't want it to get worse ...)

philipcmonk replied:
Done

philipcmonk commented:
I removed the snapshot stuff, hooked up `|rekey`, and moved `+moon` to `|moon` to make it stateful.  I also added `|moon-cycle-keys` and `|moon-breach` to let you change the keys of a moon or breach it.

joemfb commented on pkg/arvo/gen/hood/moon.hoon:
```diff
@@ -0,0 +1,44 @@
+=/  =pass
+  ?.  =(*pass public-key)
+    public-key
+  =/  cub  (pit:nu:crub:crypto 128 (sham mon life=1 eny))
```
I'm slightly partial to using 512 bits of entropy ("more is better"). And it seems a little strange to to take a 128-bit hash, and then take a 512-bit hash of it (in `+pit:nu:crub`). But I'm also not sure it's an actual problem. Is this just to align 128-bit addresses with 128-bit seeds?

philipcmonk replied:
nah, it's just the only hash function that takes a noun instead of an atom.  I've replaced it with `(shaz (jam mon life eny))`, which should give us 512 bits.

joemfb commented on pkg/arvo/lib/hood/helm.hoon:
```diff
@@ -79,14 +80,14 @@
-  =/  lyf=life  .^(@ud j+/(scot %p our)/life/(scot %da now)/(scot %p our))
-  ?.  =(+(lyf) lyf.u.sed)
-    ~&  [%wrong-private-key-life expected=+(lyf) actual=lyf.u.sed]
+  (emit %rekey / lyf.u.sed key.u.sed)
```
Much obliged!

joemfb commented on pkg/arvo/lib/ph/azimuth.hoon:
```diff
@@ -183,31 +248,68 @@
+  ?.  =(%earl (clan:title who))
+    (raw-ship:(ph-tests our) who `(dawn who ~))
+  =/  spon=ship  (^sein:title who)
+  =/  cub  (pit:nu:crub:crypto 128 (sham who life=1 %entropy))
```
Also 128-bits here, although I'm not sure that matters.

philipcmonk replied:
it's just a test, but may as well.

joemfb commented on pkg/arvo/lib/tapp.hoon:
```diff
@@ -254,10 +254,10 @@
+      ~&  [%tapp-reset dap.bowl]
+      `this-tapp
+      ::  ~|  [%tapp-load-incompatible dap.bowl]
+      ::  !!
```
I think this should remain a crash on `master`. It can be disabled for development if necessary, but we don't want to accidentally OTA with the state-discarding reset. Tapps that need to migrate state should be able to use `$&`.

philipcmonk replied:
Ah, yes.
